File: framec_tests/golang/hierarchical_guard/hierarchical_guard.go
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

package hierarchical_guard

import "golang/framelang"


func NewHierarchicalGuard() HierarchicalGuard {
    m := &hierarchicalGuardStruct{}
    
    // Validate interfaces
    var _ HierarchicalGuard = m
    var _ HierarchicalGuard_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewHierarchicalGuardCompartment(HierarchicalGuardState_I)
    
    // Override domain variables.
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type HierarchicalGuardState uint

const (
    HierarchicalGuardState_I HierarchicalGuardState = iota
    HierarchicalGuardState_S
    HierarchicalGuardState_S0
    HierarchicalGuardState_S1
    HierarchicalGuardState_S2
    HierarchicalGuardState_S3
    HierarchicalGuardState_S4
)

type HierarchicalGuard interface {
    A(i int) 
    B(i int) 
}

type HierarchicalGuard_actions interface {
    log(msg string) 
}


type hierarchicalGuardStruct struct {
    _compartment_ *HierarchicalGuardCompartment
    _nextCompartment_ *HierarchicalGuardCompartment
    tape []string
}

//===================== Interface Block ===================//

func (m *hierarchicalGuardStruct) A(i int)  {
    params := make(map[string]interface{})
    params["i"] = i
    e := framelang.FrameEvent{Msg:"A", Params:params}
    m._mux_(&e)
}

func (m *hierarchicalGuardStruct) B(i int)  {
    params := make(map[string]interface{})
    params["i"] = i
    e := framelang.FrameEvent{Msg:"B", Params:params}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *hierarchicalGuardStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case HierarchicalGuardState_I:
        m._HierarchicalGuardState_I_(e)
    case HierarchicalGuardState_S:
        m._HierarchicalGuardState_S_(e)
    case HierarchicalGuardState_S0:
        m._HierarchicalGuardState_S0_(e)
    case HierarchicalGuardState_S1:
        m._HierarchicalGuardState_S1_(e)
    case HierarchicalGuardState_S2:
        m._HierarchicalGuardState_S2_(e)
    case HierarchicalGuardState_S3:
        m._HierarchicalGuardState_S3_(e)
    case HierarchicalGuardState_S4:
        m._HierarchicalGuardState_S4_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *hierarchicalGuardStruct) _HierarchicalGuardState_I_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S)
        m._transition_(compartment)
        
        return
    }
}

func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        m.log("S.A")
        if e.Params["i"].(int) < 10 {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S0)
            m._transition_(compartment)
            return
        } else {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S1)
            m._transition_(compartment)
            return
        }
        
        return
    case "B":
        m.log("S.B")
        if e.Params["i"].(int) < 10 {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S2)
            m._transition_(compartment)
            return
        } else {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S3)
            m._transition_(compartment)
            return
        }
        
        return
    }
}

func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S0_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        m.log("S0.A")
        if e.Params["i"].(int) > 0 {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S2)
            m._transition_(compartment)
            return
        } else {
        }
        
      //  fall through else branch
    case "B":
        m.log("S0.B")
        if e.Params["i"].(int) > 0 {
        } else {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S1)
            m._transition_(compartment)
            return
        }
        
    }
    m._HierarchicalGuardState_S_(e)
    
}  //  fall through then branch


func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S1_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        m.log("S1.A")
        if e.Params["i"].(int) > 5 {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S3)
            m._transition_(compartment)
            return
        } else {
        }
        
    }
    m._HierarchicalGuardState_S0_(e)
    
}  //  fall through else branch


func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S2_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        m.log("S2.A")
        if e.Params["i"].(int) > 10 {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S4)
            m._transition_(compartment)
            return
        } else {
        }
        
      //  fall through then branch
    case "B":
        m.log("S2.B")
        if !(e.Params["i"].(int) > 10) {
        } else {
            compartment := NewHierarchicalGuardCompartment(HierarchicalGuardState_S4)
            m._transition_(compartment)
            return
        }
        
    }
    m._HierarchicalGuardState_S1_(e)
    
}  //  fall through then branch


func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S3_(e *framelang.FrameEvent) {
    switch e.Msg {
    case "A":
        m.log("S3.A")
        if e.Params["i"].(int) > 0 {
            m.log("stop")
            
            return
        } else {
            m.log("continue")
        }
        
    case "B":
        m.log("S3.B")
        if e.Params["i"].(int) > 0 {
            m.log("continue")
        } else {
            m.log("stop")
            
            return
        }
        
    }
    m._HierarchicalGuardState_S_(e)
    
}

func (m *hierarchicalGuardStruct) _HierarchicalGuardState_S4_(e *framelang.FrameEvent) {
    switch e.Msg {
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *hierarchicalGuardStruct) _transition_(compartment *HierarchicalGuardCompartment) {
    m._nextCompartment_ = compartment
}

func (m *hierarchicalGuardStruct) _do_transition_(nextCompartment *HierarchicalGuardCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *hierarchicalGuardStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type HierarchicalGuardCompartment struct {
    State HierarchicalGuardState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewHierarchicalGuardCompartment(state HierarchicalGuardState) *HierarchicalGuardCompartment {
    c := &HierarchicalGuardCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/hierarchical_guard/hierarchical_guard.py
#Emitted from framec_v0.11.0



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class HierarchicalGuard:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_I', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        
        # Initialize domain
        
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def A(self,i: int):
        parameters = {}
        parameters["i"] = i
        __e = FrameEvent("A",parameters)
        self.__kernel(__e)
    
    def B(self,i: int):
        parameters = {}
        parameters["i"] = i
        __e = FrameEvent("B",parameters)
        self.__kernel(__e)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $I
    
    def __hierarchicalguard_state_I(self, __e, compartment):
        if __e._message == ">":
            next_compartment = None
            next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $S
    
    def __hierarchicalguard_state_S(self, __e, compartment):
        if __e._message == "A":
            self.log_do("S.A")
            if  __e._parameters["i"] < 10:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S0', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S0', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S1', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
        elif __e._message == "B":
            self.log_do("S.B")
            if  __e._parameters["i"] < 10:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S0', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S1', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S2', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S3', next_compartment)
                self.__transition(next_compartment)
                return
            
            return
    
    # ----------------------------------------
    # $S0
    
    def __hierarchicalguard_state_S0(self, __e, compartment):
        if __e._message == "A":
            self.log_do("S0.A")
            if  __e._parameters["i"] > 0:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S0', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S1', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S2', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                pass
            
          #  fall through else branch
        elif __e._message == "B":
            self.log_do("S0.B")
            if  __e._parameters["i"] > 0:
                pass
            else:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S0', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S1', next_compartment)
                self.__transition(next_compartment)
                return
            
        
        self.__hierarchicalguard_state_S(__e, compartment.parent_compartment)
        
      #  fall through then branch
    
    
    # ----------------------------------------
    # $S1
    
    def __hierarchicalguard_state_S1(self, __e, compartment):
        if __e._message == "A":
            self.log_do("S1.A")
            if  __e._parameters["i"] > 5:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S', next_compartment)
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S3', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                pass
            
        
        self.__hierarchicalguard_state_S0(__e, compartment.parent_compartment)
        
      #  fall through else branch
    
    
    # ----------------------------------------
    # $S2
    
    def __hierarchicalguard_state_S2(self, __e, compartment):
        if __e._message == "A":
            self.log_do("S2.A")
            if  __e._parameters["i"] > 10:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S4', next_compartment)
                self.__transition(next_compartment)
                return
            else:
                pass
            
          #  fall through then branch
        elif __e._message == "B":
            self.log_do("S2.B")
            if  not (__e._parameters["i"] > 10):
                pass
            else:
                next_compartment = None
                next_compartment = HierarchicalGuardCompartment('__hierarchicalguard_state_S4', next_compartment)
                self.__transition(next_compartment)
                return
            
        
        self.__hierarchicalguard_state_S1(__e, compartment.parent_compartment)
        
      #  fall through then branch
    
    
    # ----------------------------------------
    # $S3
    
    def __hierarchicalguard_state_S3(self, __e, compartment):
        if __e._message == "A":
            self.log_do("S3.A")
            if  __e._parameters["i"] > 0:
                self.log_do("stop")
                
                return
            else:
                self.log_do("continue")
            
        elif __e._message == "B":
            self.log_do("S3.B")
            if  __e._parameters["i"] > 0:
                self.log_do("continue")
            else:
                self.log_do("stop")
                
                return
            
        
        self.__hierarchicalguard_state_S(__e, compartment.parent_compartment)
        
    
    # ----------------------------------------
    # $S4
    
    def __hierarchicalguard_state_S4(self, __e, compartment):
        pass
        
    
    # ===================== Actions Block =================== #
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__hierarchicalguard_state_I':
            self.__hierarchicalguard_state_I(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S':
            self.__hierarchicalguard_state_S(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S0':
            self.__hierarchicalguard_state_S0(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S1':
            self.__hierarchicalguard_state_S1(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S2':
            self.__hierarchicalguard_state_S2(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S3':
            self.__hierarchicalguard_state_S3(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchicalguard_state_S4':
            self.__hierarchicalguard_state_S4(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class HierarchicalGuardCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    